From 9d5c2ee3a7bb7c86a9fa243fc1f0760a68a23057 Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Wed, 13 Mar 2024 05:00:47 +0200
Subject: [PATCH] [SW-402] fs1: status: add status update

This patch adds functionality which allows boot rom to log
its progress in status registers which are part of
security block.

Change-Id: I0fdbbd6aad8cd027e131e0c96eb9172fce511783
---

diff --git a/arch/fs1/arc/common.c b/arch/fs1/arc/common.c
index 831e6b6..4d48b14 100644
--- a/arch/fs1/arc/common.c
+++ b/arch/fs1/arc/common.c
@@ -9,6 +9,7 @@
 #include <fs1_regs.h>
 #include <fs1_reg_map.h>
 #include <fs1_reg_mask.h>
+#include <status.h>
 
 uint8_t boot_rom_get_die_id(void)
 {
@@ -96,3 +97,8 @@
 	if (boot_rom_global.die_type != IO_DIE_MASTER)
 		regmap_writel(0, IO0_H0_PSOC_GLOBAL_CONF(mmGLOBAL_CONF_SCRATCHPAD_2));
 }
+
+uint32_t boot_rom_get_status_base(void)
+{
+	return IO0_HALF0_PSOC_BOOT_SECURITY(mmPSOC_SECURITY_SW_STATUS_0);
+}
diff --git a/arch/fs1/arc/progress_bits.h b/arch/fs1/arc/progress_bits.h
new file mode 100644
index 0000000..3efae37
--- /dev/null
+++ b/arch/fs1/arc/progress_bits.h
@@ -0,0 +1,129 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2024 HabanaLabs Ltd.
+ */
+
+#ifndef _PROGRESS_BITS_H_
+#define _PROGRESS_BITS_H_
+
+enum boot_rom_progress {
+	PROGRESS_INIT = 0,
+	PROGRESS_INIT_ARC1_ICCM_INIT_FAILED = 1,
+	PROGRESS_INIT_SPI_BOOT_DESC_EXCEEDS_MAX_SIZE = 2,
+	PROGRESS_INIT_SPI_BOOT_DESC_DMA_START = 3,
+	PROGRESS_INIT_SPI_BOOT_DESC_CRC_FAIL = 4,
+	PROGRESS_INIT_SPI_BOOT_DESC_UNALIGNED_DEST = 5,
+	PROGRESS_INIT_SPI_BOOT_DESC_INVALID_IMG_SIZE = 6,
+	PROGRESS_INIT_SPI_BOOT_DESC_UNALIGNED_IMG_SIZE = 7,
+	PROGRESS_INIT_SPI_BOOT_DESC_DMA_ERROR = 8,
+	PROGRESS_INIT_SPI_BOOT_DESC_DMA_FINISH = 9,
+	PROGRESS_INIT_SPI_BOOT_DESC_HDR_MAGIC_FAIL = 10,
+	PROGRESS_INIT_SPI_BOOT_DESC_INVALID_IMG_OFFSET = 11,
+	PROGRESS_INIT_SPI_BOOT_DESC_VERIFICATION_FAIL = 12,
+	PROGRESS_INIT_SPI_BOOT_DESC_INFO_SET_FAIL = 13,
+	PROGRESS_INIT_SPI_BOOT_DESC_DATA_SET_FAIL = 14,
+	PROGRESS_INIT_SPI_BOOT_DESC_INFO_GET_FAIL = 15,
+	PROGRESS_INIT_SPI_BOOT_DESC_DATA_GET_FAIL = 16,
+	PROGRESS_INIT_SPI_BOOT_DESC_FAIL = 17,
+	PROGRESS_INIT_EFUSE_INIT_FAIL = 18,
+	PROGRESS_INIT_EFUSE_SPI = 19,
+	PROGRESS_INIT_EFUSE_SPI_SLV_TIMEDOUT = 20,
+	PROGRESS_REVOKE_LOC = 21,
+	PROGRESS_REVOKE_LOC_0 = 22,
+	PROGRESS_REVOKE_LOC_1 = 23,
+	PROGRESS_KEY_LOC = 24,
+	PROGRESS_KEY_LOC_0 = 25,
+	PROGRESS_KEY_LOC_1 = 26,
+	PROGRESS_VER_LOC = 27,
+	PROGRESS_VER_LOC_0 = 28,
+	PROGRESS_VER_LOC_1 = 29,
+	PROGRESS_SLV_DIE_WAITING_ON_RECOVERY_IND = 30,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_START = 31,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_VERSION_FAIL = 32,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_KEY_REVOKED = 33,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 34,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_AUTH_FAIL = 35,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_SUCCESS = 36,
+	PROGRESS_BRINGUP_PCI_PRI = 37,
+	PROGRESS_BRINGUP_PCI_PRI_LOADADDR_INVALID = 38,
+	PROGRESS_BRINGUP_PCI_PRI_STARTADDR_INVALID = 39,
+	PROGRESS_BRINGUP_PCI_PRI_TOO_BIG = 40,
+	PROGRESS_BRINGUP_PCI_PRI_CRC_FAIL = 41,
+	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_DEST_ADDR = 42,
+	PROGRESS_BRINGUP_PCI_PRI_INVALID_IMG_SIZE = 43,
+	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_IMG_SIZE = 44,
+	PROGRESS_BRINGUP_PCI_PRI_DMA_ERROR = 45,
+	PROGRESS_BRINGUP_PCI_PRI_DMA = 46,
+	PROGRESS_BRINGUP_PCI_PRI_KEY_REVOKED = 47,
+	PROGRESS_BRINGUP_PCI_PRI_KEY_AUTH_FAIL = 48,
+	PROGRESS_BRINGUP_PCI_PRI_SECURITY = 49,
+	PROGRESS_BRINGUP_PCI_SEC = 50,
+	PROGRESS_BRINGUP_PCI_SEC_LOADADDR_INVALID = 51,
+	PROGRESS_BRINGUP_PCI_SEC_STARTADDR_INVALID = 52,
+	PROGRESS_BRINGUP_PCI_SEC_TOO_BIG = 53,
+	PROGRESS_BRINGUP_PCI_SEC_CRC_FAIL = 54,
+	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_DEST_ADDR = 55,
+	PROGRESS_BRINGUP_PCI_SEC_INVALID_IMG_SIZE = 56,
+	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_IMG_SIZE = 57,
+	PROGRESS_BRINGUP_PCI_SEC_DMA_ERROR = 58,
+	PROGRESS_BRINGUP_PCI_SEC_DMA = 59,
+	PROGRESS_BRINGUP_PCI_SEC_KEY_REVOKED = 60,
+	PROGRESS_BRINGUP_PCI_SEC_KEY_AUTH_FAIL = 61,
+	PROGRESS_BRINGUP_PCI_SEC_SECURITY = 62,
+	PROGRESS_BOOT_ROM_TPM = 63,
+	PROGRESS_BOOT_ROM_TPM_INIT_FAIL = 64,
+	PROGRESS_BOOT_ROM_TPM_STARTUP_FAIL = 65,
+	PROGRESS_BOOT_ROM_TPM_PCR_EXTEND_FAIL = 67,
+	PROGRESS_BOOT_ROM_TPM_FINISH = 68,
+	PROGRESS_PPBOOT_START = 69,
+	PROGRESS_PPBOOT_FINISH = 70,
+	PROGRESS_BOOT_ROM_IN_RECOVERY = 71,
+	PROGRESS_BOOT_ROM_IN_ROM_RECOVERY = 72,
+	PROGRESS_BOOT_ROM_PCIE_LINK_START = 73,
+	PROGRESS_BOOT_ROM_PCIE_LINK_INIT = 74,
+	PROGRESS_BOOT_ROM_PCIE_PHY_SRAM_INIT_TIMEDOUT = 75,
+	PROGRESS_BOOT_ROM_PCIE_PHY_PARA_STS_CLR_TIMEDOUT = 76,
+	PROGRESS_BOOT_ROM_PCIE_LINK_LIMIT_GEN1 = 77,
+	PROGRESS_BOOT_ROM_PCIE_LINK_IATU_CONFIG = 78,
+	PROGRESS_BOOT_ROM_PCIE_LINK_FINISH = 79,
+	PROGRESS_BOOT_ROM_D2D_LINK_START = 80,
+	PROGRESS_BOOT_ROM_D2D_LINK_STS_BUSY = 81,
+	PROGRESS_BOOT_ROM_D2D_LINK_INTER_DIE_SYNC = 82,
+	PROGRESS_BOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 83,
+	PROGRESS_BOOT_ROM_D2D_LINK_INIT_SUCCESS = 84,
+	PROGRESS_BOOT_ROM_RECOVERY_END = 85,
+	PROGRESS_BOOT_ROM_IN_RECOVERY_IMG = 86,
+	PROGRESS_INIT_RECOVERY_IMG_BOOT_DESC_EXCEEDS_MAX_SIZE = 87,
+	PROGRESS_INIT_RECOVERY_IMG_BOOT_DESC_DMA_START = 88,
+	PROGRESS_INIT_RECOVERY_IMG_BOOT_DESC_CRC_FAIL = 89,
+	PROGRESS_INIT_RECOVERY_IMG_BOOT_DESC_UNALIGNED_DEST = 90,
+	PROGRESS_INIT_RECOVERY_IMG_BOOT_DESC_INVALID_IMG_SIZE = 91,
+	PROGRESS_INIT_RECOVERY_IMG_BOOT_DESC_UNALIGNED_IMG_SIZE = 92,
+	PROGRESS_INIT_RECOVERY_IMG_BOOT_DESC_DMA_ERROR = 93,
+	PROGRESS_INIT_RECOVERY_IMG_BOOT_DESC_DMA_FINISH = 94,
+	PROGRESS_INIT_RECOVERY_IMG_BOOT_DESC_HDR_MAGIC_FAIL = 95,
+	PROGRESS_INIT_RECOVERY_IMG_BOOT_DESC_INVALID_IMG_OFFSET = 96,
+	PROGRESS_INIT_RECOVERY_IMG_BOOT_DESC_VERIFICATION_FAIL = 97,
+	PROGRESS_INIT_RECOVERY_IMG_BOOT_DESC_FAIL = 98,
+	PROGRESS_RECOVERY_IMG_BOOT_DESC_VERIFY_START = 99,
+	PROGRESS_RECOVERY_IMG_BOOT_DESC_VERIFY_VERSION_FAIL = 100,
+	PROGRESS_RECOVERY_IMG_BOOT_DESC_VERIFY_KEY_REVOKED = 101,
+	PROGRESS_RECOVERY_IMG_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 102,
+	PROGRESS_RECOVERY_IMG_BOOT_DESC_VERIFY_AUTH_FAIL = 103,
+	PROGRESS_RECOVERY_IMG_BOOT_DESC_VERIFY_SUCCESS = 104,
+	PROGRESS_RECOVERY_IMG_IMG0 = 105,
+	PROGRESS_RECOVERY_IMG_IMG0_LOADADDR_INVALID = 106,
+	PROGRESS_BOOT_REOCVERY_IMG_IMG0_STARTADDR_INVALID = 107,
+	PROGRESS_BOOT_REOCVERY_IMG_IMG0_DMA_START = 108,
+	PROGRESS_BOOT_REOCVERY_IMG_IMG0_DMA_CRC_FAILED = 109,
+	PROGRESS_BOOT_REOCVERY_IMG_IMG0_DMA_UNALIGNED_DEST = 110,
+	PROGRESS_BOOT_REOCVERY_IMG_IMG0_DMA_INVALID_IMG_SIZE = 111,
+	PROGRESS_BOOT_REOCVERY_IMG_IMG0_DMA_UNALIGNED_IMG_SIZE = 112,
+	PROGRESS_BOOT_REOCVERY_IMG_IMG0_DMA_ERROR = 113,
+	PROGRESS_BOOT_REOCVERY_IMG_IMG0_DMA_FINISH = 114,
+	PROGRESS_BOOT_REOCVERY_IMG_IMG0_VERIFICATION_FAIL = 115,
+	PROGRESS_BOOT_REOCVERY_IMG_IMG0_START = 116,
+	PROGRESS_BOOT_REOCVERY_IMG_IMG0_FINISH = 117,
+};
+
+#endif /* _PROGRESS_BITS_H_ */
diff --git a/common/common.c b/common/common.c
index 6053351..8f4dfe9 100644
--- a/common/common.c
+++ b/common/common.c
@@ -5,6 +5,8 @@
  */
 
 #include <boot_rom.h>
+#include <status.h>
+#include <common.h>
 
 void boot_rom_init(void)
 {
@@ -18,4 +20,6 @@
 
 	/* clear boot rom Status registers */
 	boot_rom_clear_status();
+
+	boot_rom_status_update(PROGRESS_INIT);
 }
diff --git a/common/status.c b/common/status.c
new file mode 100644
index 0000000..8c88ec8
--- /dev/null
+++ b/common/status.c
@@ -0,0 +1,48 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2023 HabanaLabs Ltd.
+ * All Rights Reserved.
+ */
+
+#include <boot_rom.h>
+#include <status.h>
+
+void boot_rom_status_update(enum boot_rom_progress progress)
+{
+	uint32_t status_base, reg, offset;
+
+	status_base = boot_rom_get_status_base();
+	offset = progress / BOOT_ROM_STATUS_REG_SZ_BITS;
+	progress = progress % BOOT_ROM_STATUS_REG_SZ_BITS;
+
+	reg = regmap_readl(status_base + offset * 4);
+	reg |= BIT(progress);
+	regmap_writel(reg, (status_base + offset * 4));
+}
+
+void boot_rom_status_update_bringup(enum boot_rom_progress progress, uint32_t position)
+{
+	boot_rom_status_update(progress + position);
+}
+
+void boot_rom_status_update_val(enum boot_rom_progress progress, uint32_t val, uint8_t bits)
+{
+	uint32_t status_base, reg, offset;
+
+	status_base = boot_rom_get_status_base();
+	offset = progress / BOOT_ROM_STATUS_REG_SZ_BITS;
+	progress = progress % BOOT_ROM_STATUS_REG_SZ_BITS;
+
+	reg = regmap_readl(status_base + offset * 4);
+	reg |= val << progress;
+	regmap_writel(reg, status_base + offset * 4);
+
+	if ((progress + bits) > BOOT_ROM_STATUS_REG_SZ_BITS) {
+		offset++;
+		val = val >> (BOOT_ROM_STATUS_REG_SZ_BITS - progress);
+		reg = regmap_readl(status_base + offset * 4);
+		reg |= val;
+		regmap_writel(reg, status_base + offset * 4);
+	}
+}
+
diff --git a/common/status.h b/common/status.h
new file mode 100644
index 0000000..112f470
--- /dev/null
+++ b/common/status.h
@@ -0,0 +1,75 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2023 HabanaLabs Ltd.
+ */
+
+#ifndef _STATUS_H_
+#define _STATUS_H_
+
+#include <stdint.h>
+#include <progress_bits.h>
+#include <mem_io.h>
+
+#define BOOT_ROM_STATUS_REG_SZ_BITS	32
+
+enum boot_rom_bringup_proggress {
+	PROGRESS_BRINGUP_START,
+	PROGRESS_BRINGUP_LOADADDR_INVALID,
+	PROGRESS_BRINGUP_STARTADDR_INVALID,
+	PROGRESS_BRINGUP_TOO_BIG,
+	PROGRESS_BRINGUP_CRC_FAIL,
+	PROGRESS_BRINGUP_UNALIGNED_DEST_ADDR,
+	PROGRESS_BRINGUP_INVALID_IMG_SIZE,
+	PROGRESS_BRINGUP_UNALIGNED_IMG_SIZE,
+	PROGRESS_BRINGUP_DMA_ERROR,
+	PROGRESS_BRINGUP_DMA,
+	PROGRESS_BRINGUP_KEY_REVOKED,
+	PROGRESS_BRINGUP_KEY_AUTH_FAIL,
+	PROGRESS_BRINGUP_SECURITY,
+};
+
+enum boot_rom_boot_desc_progress {
+	PROGRESS_BOOT_DESC_EXCEEDS_MAX_SIZE,
+	PROGRESS_DMA_START,
+	PROGRESS_DMA_CRC_FAILED,
+	PROGRESS_DMA_UNALIGNED_DEST,
+	PROGRESS_DMA_INVALID_IMG_SIZE,
+	PROGRESS_DMA_UNALIGNED_IMG_SIZE,
+	PROGRESS_DMA_ERROR,
+	PROGRESS_DMA_FINISH,
+	PROGRESS_BOOT_DESC_HDR_MAGIC_FAIL,
+	PROGRESS_BOOT_DESC_INVALID_IMG_OFFSET,
+	PROGRESS_BOOT_DESC_VERIFICATION_FAIL,
+};
+
+enum boot_rom_boot_desc_verify_progress {
+	PROGRESS_VERIFY_START,
+	PROGRESS_VERIFY_VERSION_FAIL,
+	PROGRESS_VERIFY_KEY_REVOKED,
+	PROGRESS_VERIFY_KEY_HASH_FAIL,
+	PROGRESS_VERIFY_AUTH_FAIL,
+	PROGRESS_VERIFY_SUCCESS,
+};
+
+enum boot_rom_erom_img_progress {
+	PROGRESS_EROM_BRINGUP_START,
+	PROGRESS_EROM_LOADADDR_INVALID,
+	PROGRESS_EROM_STARTADDR_INVALID,
+	PROGRESS_EROM_DMA_START,
+	PROGRESS_EROM_DMA_CRC_FAILED,
+	PROGRESS_EROM_DMA_UNALIGNED_DEST_ADDR,
+	PROGRESS_EROM_DMA_INVALID_IMG_SIZE,
+	PROGRESS_EROM_DMA_UNALIGNED_IMG_SIZE,
+	PROGRESS_EROM_DMA_ERROR,
+	PROGRESS_EROM_DMA_FINISH,
+	PROGRESS_EROM_IMG_VERIFICATION_FAIL,
+	PROGRESS_EROM_IMG_START,
+	PROGRESS_EROM_IMG_FINISH,
+};
+
+void boot_rom_status_update(enum boot_rom_progress progress);
+void boot_rom_status_update_bringup(enum boot_rom_progress progress, uint32_t position);
+void boot_rom_status_update_val(enum boot_rom_progress progress, uint32_t val, uint8_t bits);
+uint32_t boot_rom_get_status_base(void);
+
+#endif /* _STATUS_H_ */
